File: python-server/app/services/cooldown_service.py
"""Provider cooldown management service.

Manages temporary provider disabling based on error responses.
Uses in-memory storage with TTL-based automatic expiration.
Thread-safe implementation using locks.
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Optional
import threading
import time

from loguru import logger


@dataclass
class CooldownEntry:
    """Cooldown cache value for a provider.

    Attributes:
        provider_key: Unique provider identifier
        exception_type: Type of exception (rate_limit, auth_error, timeout, server_error)
        status_code: HTTP status code that triggered cooldown
        timestamp: Unix timestamp when cooldown started
        cooldown_time: Cooldown duration in seconds
        error_message: Optional error message for logging
    """

    provider_key: str
    exception_type: str
    status_code: int
    timestamp: float
    cooldown_time: int
    error_message: Optional[str] = None

    @property
    def expires_at(self) -> float:
        """Calculate expiration timestamp."""
        return self.timestamp + self.cooldown_time

    @property
    def is_expired(self) -> bool:
        """Check if cooldown has expired."""
        return time.time() >= self.expires_at

    @property
    def remaining_seconds(self) -> int:
        """Get remaining cooldown seconds."""
        remaining = self.expires_at - time.time()
        return max(0, int(remaining))

    @property
    def started_at_iso(self) -> str:
        """Get start time in ISO format."""
        return datetime.fromtimestamp(self.timestamp, tz=timezone.utc).isoformat()

    @property
    def expires_at_iso(self) -> str:
        """Get expiration time in ISO format."""
        return datetime.fromtimestamp(self.expires_at, tz=timezone.utc).isoformat()


@dataclass
class CooldownConfig:
    """Cooldown configuration.

    Attributes:
        enabled: Whether cooldown feature is enabled
        default_cooldown_secs: Default cooldown duration in seconds
        max_cooldown_secs: Maximum cooldown time cap
        cooldown_status_codes: Status codes that trigger cooldown
        non_cooldown_status_codes: Status codes that should NOT trigger cooldown
        cooldown_durations: Status code specific cooldown durations
    """

    enabled: bool = True
    default_cooldown_secs: int = 60
    max_cooldown_secs: int = 600

    # Status codes that trigger cooldown: only 429 (rate limit) and 5xx (server errors)
    cooldown_status_codes: set[int] = field(
        default_factory=lambda: {
            429,  # Rate limit - provider is overloaded
            500,  # Internal server error
            501,  # Not implemented
            502,  # Bad gateway
            503,  # Service unavailable
            504,  # Gateway timeout
        }
    )

    # Status codes that should NOT trigger cooldown (client errors, auth errors)
    # 4xx errors (except 429) are typically client-side issues, not provider issues
    non_cooldown_status_codes: set[int] = field(
        default_factory=lambda: {400, 401, 403, 404, 408, 422}
    )

    # Status code specific cooldown durations
    cooldown_durations: dict[int, int] = field(
        default_factory=lambda: {
            429: 60,  # Rate limit - 1 minute
            500: 30,  # Internal server error - 30 seconds
            502: 30,  # Bad gateway - 30 seconds
            503: 60,  # Service unavailable - 1 minute
            504: 30,  # Gateway timeout - 30 seconds
        }
    )


def _get_exception_type(status_code: int) -> str:
    """Determine exception type based on status code."""
    if status_code == 429:
        return "rate_limit"
    elif 500 <= status_code < 600:
        return "server_error"
    else:
        return "unknown_error"


class CooldownService:
    """Provider cooldown management service.

    Manages temporary provider disabling based on error responses.
    Uses in-memory storage with TTL-based automatic expiration.
    Thread-safe implementation using locks.
    """

    def __init__(self, config: Optional[CooldownConfig] = None):
        self._config = config or CooldownConfig()
        self._cooldowns: dict[str, CooldownEntry] = {}
        self._lock = threading.Lock()

    @property
    def config(self) -> CooldownConfig:
        """Get current cooldown configuration."""
        return self._config

    def update_config(self, config: CooldownConfig) -> None:
        """Update cooldown configuration."""
        with self._lock:
            self._config = config

    def should_trigger_cooldown(self, status_code: int) -> bool:
        """Check if a status code should trigger cooldown.

        Args:
            status_code: HTTP status code from provider response

        Returns:
            True if cooldown should be triggered
        """
        if not self._config.enabled:
            return False

        # Explicitly excluded codes (client errors)
        if status_code in self._config.non_cooldown_status_codes:
            return False

        # Explicitly included codes
        if status_code in self._config.cooldown_status_codes:
            return True

        # Default: trigger for 5xx errors
        return 500 <= status_code < 600

    def add_cooldown(
        self,
        provider_key: str,
        status_code: int,
        error_message: Optional[str] = None,
        cooldown_time: Optional[int] = None,
    ) -> CooldownEntry:
        """Add a provider to cooldown.

        Args:
            provider_key: Unique provider identifier
            status_code: HTTP status code that triggered cooldown
            error_message: Optional error message for logging
            cooldown_time: Override cooldown duration

        Returns:
            Created CooldownEntry
        """
        exception_type = _get_exception_type(status_code)

        if cooldown_time is None:
            cooldown_time = self._config.cooldown_durations.get(
                status_code, self._config.default_cooldown_secs
            )

        # Cap cooldown time
        cooldown_time = min(cooldown_time, self._config.max_cooldown_secs)

        entry = CooldownEntry(
            provider_key=provider_key,
            exception_type=exception_type,
            status_code=status_code,
            timestamp=time.time(),
            cooldown_time=cooldown_time,
            error_message=error_message,
        )

        with self._lock:
            self._cooldowns[provider_key] = entry

        logger.warning(
            f"Provider '{provider_key}' entered cooldown: "
            f"status_code={status_code}, type={exception_type}, "
            f"duration={cooldown_time}s"
        )

        return entry

    def is_in_cooldown(self, provider_key: str) -> bool:
        """Check if a provider is currently in cooldown.

        Args:
            provider_key: Unique provider identifier

        Returns:
            True if provider is in cooldown (not expired)
        """
        if not self._config.enabled:
            return False

        with self._lock:
            entry = self._cooldowns.get(provider_key)
            if entry is None:
                return False

            if entry.is_expired:
                # Auto-cleanup expired entry
                del self._cooldowns[provider_key]
                logger.info(
                    f"Provider '{provider_key}' cooldown expired, now available"
                )
                return False

            return True

    def get_cooldown(self, provider_key: str) -> Optional[CooldownEntry]:
        """Get cooldown entry for a provider.

        Args:
            provider_key: Unique provider identifier

        Returns:
            CooldownEntry if in cooldown, None otherwise
        """
        with self._lock:
            entry = self._cooldowns.get(provider_key)
            if entry is None:
                return None

            if entry.is_expired:
                del self._cooldowns[provider_key]
                return None

            return entry

    def remove_cooldown(self, provider_key: str) -> bool:
        """Manually remove a provider from cooldown.

        Args:
            provider_key: Unique provider identifier

        Returns:
            True if cooldown was removed, False if not found
        """
        with self._lock:
            if provider_key in self._cooldowns:
                del self._cooldowns[provider_key]
                logger.info(f"Provider '{provider_key}' cooldown manually removed")
                return True
            return False

    def get_all_cooldowns(self) -> dict[str, CooldownEntry]:
        """Get all active cooldowns (excluding expired).

        Returns:
            Dict of provider_key -> CooldownEntry
        """
        with self._lock:
            # Cleanup expired entries
            expired_keys = [k for k, v in self._cooldowns.items() if v.is_expired]
            for key in expired_keys:
                del self._cooldowns[key]
                logger.info(f"Provider '{key}' cooldown expired, now available")

            # Return copy to avoid modification issues
            return dict(self._cooldowns)

    def clear_all_cooldowns(self) -> int:
        """Clear all cooldowns (admin operation).

        Returns:
            Number of cooldowns cleared
        """
        with self._lock:
            count = len(self._cooldowns)
            self._cooldowns.clear()
            if count > 0:
                logger.warning(f"All cooldowns cleared by admin: {count} entries")
            return count

    def filter_available_providers(
        self,
        providers: list[Any],
        weights: list[int],
    ) -> tuple[list[Any], list[int]]:
        """Filter out providers that are in cooldown.

        Args:
            providers: List of Provider objects
            weights: Corresponding weights

        Returns:
            Tuple of (filtered_providers, filtered_weights)
        """
        if not self._config.enabled:
            return providers, weights

        # Get all cooldowns once with a single lock acquisition
        with self._lock:
            # Cleanup expired entries first
            expired_keys = [k for k, v in self._cooldowns.items() if v.is_expired]
            for key in expired_keys:
                del self._cooldowns[key]
                logger.info(f"Provider '{key}' cooldown expired, now available")

            # Take snapshot of current cooldowns
            cooldown_snapshot = dict(self._cooldowns)

        filtered_providers = []
        filtered_weights = []

        for provider, weight in zip(providers, weights):
            provider_name = getattr(provider, "name", str(provider))
            entry = cooldown_snapshot.get(provider_name)

            if entry is None:
                filtered_providers.append(provider)
                filtered_weights.append(weight)
            else:
                logger.debug(
                    f"Provider '{provider_name}' skipped (in cooldown, "
                    f"{entry.remaining_seconds}s remaining)"
                )

        return filtered_providers, filtered_weights


# Global singleton
_cooldown_service: Optional[CooldownService] = None


def get_cooldown_service() -> CooldownService:
    """Get global cooldown service instance."""
    global _cooldown_service
    if _cooldown_service is None:
        _cooldown_service = CooldownService()
    return _cooldown_service


def init_cooldown_service(config: Optional[CooldownConfig] = None) -> CooldownService:
    """Initialize cooldown service with config."""
    global _cooldown_service
    _cooldown_service = CooldownService(config)
    return _cooldown_service


def trigger_cooldown_if_needed(
    provider_key: str,
    status_code: int,
    error_message: Optional[str] = None,
) -> Optional[CooldownEntry]:
    """Trigger cooldown for provider if status code warrants it.

    This is a convenience function for use in error handling code.

    Args:
        provider_key: Provider that returned the error
        status_code: HTTP status code from provider
        error_message: Optional error message for logging

    Returns:
        CooldownEntry if cooldown was triggered, None otherwise
    """
    cooldown_svc = get_cooldown_service()

    if not cooldown_svc.should_trigger_cooldown(status_code):
        return None

    return cooldown_svc.add_cooldown(
        provider_key=provider_key,
        status_code=status_code,
        error_message=error_message,
    )

File: rust-server/src/services/cooldown_service.rs
use dashmap::DashMap;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

/// Entry representing a provider in cooldown state
#[derive(Debug, Clone)]
pub struct CooldownEntry {
    pub provider_name: String,
    pub status_code: u16,
    pub exception_type: String,
    pub started_at: Instant,
    pub cooldown_duration: Duration,
    pub error_message: Option<String>,
}

impl CooldownEntry {
    /// Check if this cooldown entry has expired
    #[inline]
    pub fn is_expired(&self) -> bool {
        self.started_at.elapsed() >= self.cooldown_duration
    }

    /// Get remaining cooldown time in seconds
    pub fn remaining_secs(&self) -> u64 {
        let elapsed = self.started_at.elapsed();
        if elapsed >= self.cooldown_duration {
            0
        } else {
            (self.cooldown_duration - elapsed).as_secs()
        }
    }
}

/// Configuration for cooldown behavior
#[derive(Debug, Clone)]
pub struct CooldownConfig {
    pub enabled: bool,
    pub default_cooldown_secs: u64,
    pub max_cooldown_secs: u64,
    /// Status codes that trigger cooldown: only 429 (rate limit) and 5xx (server errors)
    pub cooldown_status_codes: Vec<u16>,
    /// Per-status-code cooldown durations
    pub cooldown_durations: HashMap<u16, u64>,
}

impl Default for CooldownConfig {
    fn default() -> Self {
        let mut cooldown_durations = HashMap::new();
        cooldown_durations.insert(429, 60); // Rate limit - longer cooldown
        cooldown_durations.insert(500, 30);
        cooldown_durations.insert(501, 30);
        cooldown_durations.insert(502, 30);
        cooldown_durations.insert(503, 60); // Service unavailable - longer
        cooldown_durations.insert(504, 30);

        Self {
            enabled: std::env::var("COOLDOWN_ENABLED")
                .map(|v| v.to_lowercase() == "true" || v == "1")
                .unwrap_or(true),
            default_cooldown_secs: std::env::var("COOLDOWN_DEFAULT_SECS")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(30),
            max_cooldown_secs: std::env::var("COOLDOWN_MAX_SECS")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(300),
            cooldown_status_codes: vec![429, 500, 501, 502, 503, 504],
            cooldown_durations,
        }
    }
}

impl CooldownConfig {
    /// Check if a status code should trigger cooldown
    #[inline]
    pub fn should_trigger_cooldown(&self, status_code: u16) -> bool {
        if !self.enabled {
            return false;
        }
        // Only 429 and 5xx trigger cooldown
        status_code == 429 || (500..600).contains(&status_code)
    }

    /// Get cooldown duration for a specific status code
    pub fn get_cooldown_duration(&self, status_code: u16) -> Duration {
        let secs = self
            .cooldown_durations
            .get(&status_code)
            .copied()
            .unwrap_or(self.default_cooldown_secs)
            .min(self.max_cooldown_secs);
        Duration::from_secs(secs)
    }

    /// Get exception type for a status code
    pub fn get_exception_type(status_code: u16) -> &'static str {
        if status_code == 429 {
            "rate_limit"
        } else {
            "server_error"
        }
    }
}

/// Thread-safe cooldown service using DashMap for lock-free concurrent access
#[derive(Debug, Clone)]
pub struct CooldownService {
    cooldowns: Arc<DashMap<String, CooldownEntry>>,
    config: CooldownConfig,
}

impl CooldownService {
    pub fn new(config: CooldownConfig) -> Self {
        Self {
            cooldowns: Arc::new(DashMap::new()),
            config,
        }
    }

    /// Check if cooldown is enabled
    #[inline]
    pub fn is_enabled(&self) -> bool {
        self.config.enabled
    }

    /// Get the configuration
    pub fn config(&self) -> &CooldownConfig {
        &self.config
    }

    /// Add a provider to cooldown
    pub fn add_cooldown(
        &self,
        provider_name: &str,
        status_code: u16,
        error_message: Option<String>,
    ) {
        if !self.config.should_trigger_cooldown(status_code) {
            return;
        }

        let entry = CooldownEntry {
            provider_name: provider_name.to_string(),
            status_code,
            exception_type: CooldownConfig::get_exception_type(status_code).to_string(),
            started_at: Instant::now(),
            cooldown_duration: self.config.get_cooldown_duration(status_code),
            error_message,
        };

        tracing::warn!(
            provider = %provider_name,
            status_code = %status_code,
            cooldown_secs = %entry.cooldown_duration.as_secs(),
            "Provider added to cooldown"
        );

        self.cooldowns.insert(provider_name.to_string(), entry);
    }

    /// Check if a provider is currently in cooldown
    #[inline]
    pub fn is_in_cooldown(&self, provider_name: &str) -> bool {
        if !self.config.enabled {
            return false;
        }

        if let Some(entry) = self.cooldowns.get(provider_name) {
            if entry.is_expired() {
                drop(entry); // Release the reference before removing
                self.cooldowns.remove(provider_name);
                false
            } else {
                true
            }
        } else {
            false
        }
    }

    /// Filter available providers, removing those in cooldown
    /// Returns (available_providers, cooled_down_count)
    ///
    /// Performance: Single iteration with lazy expiration cleanup
    pub fn filter_available_providers<'a>(
        &self,
        providers: &'a [String],
    ) -> (Vec<&'a String>, usize) {
        if !self.config.enabled {
            return (providers.iter().collect(), 0);
        }

        let mut available = Vec::with_capacity(providers.len());
        let mut cooled_down = 0;
        let mut expired_keys = Vec::new();

        for provider in providers {
            if let Some(entry) = self.cooldowns.get(provider) {
                if entry.is_expired() {
                    expired_keys.push(provider.clone());
                    available.push(provider);
                } else {
                    cooled_down += 1;
                }
            } else {
                available.push(provider);
            }
        }

        // Cleanup expired entries
        for key in expired_keys {
            self.cooldowns.remove(&key);
        }

        (available, cooled_down)
    }

    /// Get cooldown entry for a specific provider
    pub fn get_cooldown(&self, provider_name: &str) -> Option<CooldownEntry> {
        self.cooldowns.get(provider_name).map(|r| r.clone())
    }

    /// Remove a provider from cooldown manually
    pub fn remove_cooldown(&self, provider_name: &str) -> bool {
        self.cooldowns.remove(provider_name).is_some()
    }

    /// Get all current cooldowns (excluding expired)
    pub fn get_all_cooldowns(&self) -> Vec<CooldownEntry> {
        let mut result = Vec::new();
        let mut expired_keys = Vec::new();

        for entry in self.cooldowns.iter() {
            if entry.is_expired() {
                expired_keys.push(entry.key().clone());
            } else {
                result.push(entry.clone());
            }
        }

        // Cleanup expired entries
        for key in expired_keys {
            self.cooldowns.remove(&key);
        }

        result
    }

    /// Clear all cooldowns
    pub fn clear_all(&self) {
        self.cooldowns.clear();
    }

    /// Get count of active cooldowns
    pub fn active_count(&self) -> usize {
        self.cleanup_expired();
        self.cooldowns.len()
    }

    /// Cleanup expired entries
    fn cleanup_expired(&self) {
        self.cooldowns.retain(|_, entry| !entry.is_expired());
    }
}

impl Default for CooldownService {
    fn default() -> Self {
        Self::new(CooldownConfig::default())
    }
}

// Global singleton using once_cell
use once_cell::sync::Lazy;

static COOLDOWN_SERVICE: Lazy<CooldownService> = Lazy::new(CooldownService::default);

/// Get reference to the global cooldown service
/// Performance: No clone, no lock - just a static reference
#[inline]
pub fn get_cooldown_service() -> &'static CooldownService {
    &COOLDOWN_SERVICE
}

/// Reset the global cooldown service (for testing)
/// Note: This clears the DashMap but doesn't recreate the service
#[cfg(test)]
pub fn reset_cooldown_service() {
    COOLDOWN_SERVICE.clear_all();
}

/// Trigger cooldown if the status code warrants it
/// This is the main entry point for triggering cooldowns from error handlers
#[inline]
pub fn trigger_cooldown_if_needed(
    provider_name: &str,
    status_code: u16,
    error_message: Option<String>,
) {
    get_cooldown_service().add_cooldown(provider_name, status_code, error_message);
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread::sleep;

    #[test]
    fn test_cooldown_config_default() {
        let config = CooldownConfig::default();
        assert!(config.enabled);
        assert_eq!(config.default_cooldown_secs, 30);
        assert_eq!(config.max_cooldown_secs, 300);
    }

    #[test]
    fn test_should_trigger_cooldown() {
        let config = CooldownConfig::default();

        // Should trigger
        assert!(config.should_trigger_cooldown(429));
        assert!(config.should_trigger_cooldown(500));
        assert!(config.should_trigger_cooldown(502));
        assert!(config.should_trigger_cooldown(503));

        // Should NOT trigger (4xx client errors)
        assert!(!config.should_trigger_cooldown(400));
        assert!(!config.should_trigger_cooldown(401));
        assert!(!config.should_trigger_cooldown(403));
        assert!(!config.should_trigger_cooldown(404));
        assert!(!config.should_trigger_cooldown(422));
    }

    #[test]
    fn test_add_and_check_cooldown() {
        let service = CooldownService::default();

        assert!(!service.is_in_cooldown("test-provider"));

        service.add_cooldown("test-provider", 429, Some("Rate limited".to_string()));

        assert!(service.is_in_cooldown("test-provider"));
    }

    #[test]
    fn test_cooldown_expiration() {
        let mut config = CooldownConfig::default();
        config.cooldown_durations.insert(429, 1); // 1 second cooldown
        let service = CooldownService::new(config);

        service.add_cooldown("test-provider", 429, None);
        assert!(service.is_in_cooldown("test-provider"));

        sleep(Duration::from_millis(1100));
        assert!(!service.is_in_cooldown("test-provider"));
    }

    #[test]
    fn test_filter_available_providers() {
        let service = CooldownService::default();

        let providers = vec![
            "provider-a".to_string(),
            "provider-b".to_string(),
            "provider-c".to_string(),
        ];

        service.add_cooldown("provider-b", 500, None);

        let (available, cooled_down) = service.filter_available_providers(&providers);

        assert_eq!(available.len(), 2);
        assert_eq!(cooled_down, 1);
        assert!(available.contains(&&"provider-a".to_string()));
        assert!(available.contains(&&"provider-c".to_string()));
        assert!(!available.contains(&&"provider-b".to_string()));
    }

    #[test]
    fn test_remove_cooldown() {
        let service = CooldownService::default();

        service.add_cooldown("test-provider", 429, None);
        assert!(service.is_in_cooldown("test-provider"));

        assert!(service.remove_cooldown("test-provider"));
        assert!(!service.is_in_cooldown("test-provider"));

        assert!(!service.remove_cooldown("non-existent"));
    }

    #[test]
    fn test_clear_all() {
        let service = CooldownService::default();

        service.add_cooldown("provider-a", 429, None);
        service.add_cooldown("provider-b", 500, None);

        assert_eq!(service.active_count(), 2);

        service.clear_all();

        assert_eq!(service.active_count(), 0);
    }

    #[test]
    fn test_4xx_does_not_trigger_cooldown() {
        let service = CooldownService::default();

        // These should NOT trigger cooldown
        service.add_cooldown("provider-400", 400, None);
        service.add_cooldown("provider-401", 401, None);
        service.add_cooldown("provider-403", 403, None);
        service.add_cooldown("provider-404", 404, None);
        service.add_cooldown("provider-422", 422, None);

        assert!(!service.is_in_cooldown("provider-400"));
        assert!(!service.is_in_cooldown("provider-401"));
        assert!(!service.is_in_cooldown("provider-403"));
        assert!(!service.is_in_cooldown("provider-404"));
        assert!(!service.is_in_cooldown("provider-422"));
    }

    #[test]
    fn test_get_all_cooldowns() {
        let service = CooldownService::default();

        service.add_cooldown("provider-a", 429, Some("Rate limit".to_string()));
        service.add_cooldown("provider-b", 500, Some("Server error".to_string()));

        let cooldowns = service.get_all_cooldowns();

        assert_eq!(cooldowns.len(), 2);
    }

    #[test]
    fn test_exception_type() {
        assert_eq!(CooldownConfig::get_exception_type(429), "rate_limit");
        assert_eq!(CooldownConfig::get_exception_type(500), "server_error");
        assert_eq!(CooldownConfig::get_exception_type(503), "server_error");
    }

    #[test]
    fn test_remaining_secs() {
        let entry = CooldownEntry {
            provider_name: "test".to_string(),
            status_code: 429,
            exception_type: "rate_limit".to_string(),
            started_at: Instant::now(),
            cooldown_duration: Duration::from_secs(60),
            error_message: None,
        };

        assert!(entry.remaining_secs() > 55);
        assert!(entry.remaining_secs() <= 60);
    }
}
